stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,602 @@
+/*
+ * Copyright (c) 2014-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.stetho.inspector.elements;
+
+import android.os.SystemClock;
+import com.facebook.stetho.common.Accumulator;
+import com.facebook.stetho.common.ArrayListAccumulator;
+import com.facebook.stetho.common.LogUtil;
+import com.facebook.stetho.common.ThreadBound;
+import com.facebook.stetho.common.UncheckedCallable;
+import com.facebook.stetho.inspector.helper.ObjectIdMapper;
+import com.facebook.stetho.inspector.helper.ThreadBoundProxy;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Queue;
+import java.util.regex.Pattern;
+
+public final class Document extends ThreadBoundProxy {
+  private final DocumentProvider.Factory mFactory;
+
+  private DocumentProvider mDocumentProvider;
+  private ShadowDocument mShadowDOM;
+  private ObjectIdMapper mObjectIdMapper;
+
+  private UpdateListenerCollection mUpdateListeners;
+
+  private ChildEventingList mCachedChildEventingList;
+  private ArrayListAccumulator<Object> mCachedChildrenAccumulator;
+
+  private int mReferenceCounter;
+
+  public Document(DocumentProvider.Factory factory) {
+    super(factory);
+
+    mFactory = factory;
+    mObjectIdMapper = new DOMObjectIdMapper();
+    mReferenceCounter = 0;
+    mUpdateListeners = new UpdateListenerCollection();
+  }
+
+  public synchronized void addRef() {
+    if (mReferenceCounter++ == 0) {
+      mDocumentProvider = mFactory.create();
+    }
+  }
+
+  public synchronized void release() {
+    if (mReferenceCounter > 0) {
+      if (--mReferenceCounter == 0) {
+        cleanUp();
+      }
+    }
+  }
+
+  private void cleanUp() {
+    mDocumentProvider.postAndWait(new Runnable() {
+      @Override
+      public void run() {
+        mDocumentProvider.setListener(null);
+        mShadowDOM = null;
+        mObjectIdMapper.clear();
+        mDocumentProvider.dispose();
+        mUpdateListeners.clear();
+        mDocumentProvider = null;
+      }
+    });
+  }
+
+  public void addUpdateListener(UpdateListener updateListener) {
```

Comment:
We need to figure out the right way to make sure this add/remove listener stuff is thread safe. Notice how we call things like `Document.removeUpdateListener` during `DOM`'s `lastPeerUnregistered`, which will happen on a different thread.